### PR TITLE
Convert  to Poetry

### LINKED ISSUE #51: Convert to Poetry
We all know why. See pydap-extras for a good example of how to handle this stuff.

### PR DESCRIPTION
**Pull request by rod-glover**
Resolves #51 

This PR
- Changes dependency management from Pipenv to Poetry
- Eliminates unnecessary explicit dependencies 
- Upgrades many dependencies to their latest versions, most importantly `pycds`, `modelmeta`, `pydap-extras`, `sqlalchemy`
- Adjusts queries and tests accordingly
  - some queries must be more explicit about joins
  -  pycds test infrastructure now migrates test database in order to install non-table items such as functions and views
- Loosens many dependency version constraints
- Removes now superfluous Docker test infrastructure
- Moves pytest.ini into pyproject.toml
- Sets us up for #50

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    "Topic :: Internet :: WWW/HTTP :: WSGI",
+    "Topic :: Scientific/Engineering",
+    "Topic :: Software Development :: Libraries :: Python Modules",
+]
+packages = [
+    {include = "pdp_util"},
+]
+include = [
+    "pdp_util/data/alpha.png",
+    "pdp_util/data/*.css",
```
Comment (jameshiebert) [inline pyproject.toml:29]: See [this comment](https://github.com/pacificclimate/pdp_util/issues/47#issuecomment-1834385153), but this can be deferred to another PR.

File: tests/conftest.py [hunk 2]
```diff
@@ -30,52 +36,187 @@
 
 
 @pytest.fixture(scope="session")
-def engine():
-    """Test-session-wide database engine"""
+def pkg_file_root():
```
Comment (jameshiebert) [inline tests/conftest.py:39]: That's a very elegant solution to the `pkg_resources` migration! We *should* eventually try to steer away from using file system paths, and use in memory package resources if possible. But this transitions away from the deprecated behaviour very nicely.

File: tests/test_map.py [hunk 3]
```diff
@@ -1,28 +1,30 @@
...
-def test_can_instantiate():
+
+def test_can_instantiate(kwargs):
     map_app = MapApp(**kwargs)
     assert isinstance(map_app, MapApp)
 
 
 # FIXME: we can't actually _do_ a package-independent render now that the templates are in pdp... *grumble*
-def notest_minimal_render():
+def notest_minimal_render(kwargs):
```
Comment (jameshiebert) [inline tests/test_map.py:27]: This test can go away completely when #47 is closed.

**Review by jameshiebert — COMMENTED** [review]
Overall looks great modulo a few minor comments!

File: .github/workflows/black.yml [hunk 4]
```diff
@@ -9,8 +8,12 @@ jobs:
     runs-on: ubuntu-20.04
 
     steps:
-      - uses: actions/checkout@v2
-      - uses: actions/setup-python@v2
-      - uses: psf/black@stable
-        with:
-          black_args: ". --check"
\ No newline at end of file
+      - name: Temporarily disabled
```
[the remark this answers is not in the document]
Comment (rod-glover) [reply .github/workflows/black.yml:11]: Because this repo will be quite heavily changed by Black, and I wanted to defer that step to a separate PR to avoid overcomplicating this one. Should add a ref to the issue for this.

File: .github/workflows/pypi-publish.yml [hunk 5]
```diff
@@ -11,19 +11,42 @@ jobs:
...
+          # Private repo name is "pcic". We could factor that out as an env var...
+          poetry config repositories.pcic https://pypi.pacificclimate.org/
+          poetry config http-basic.pcic $PCIC_PYPI_USERNAME $PCIC_PYPI_PASSWORD
+          
+          # Pre-install initial packages (pip, setuptools, numpy) 
+          poetry install --only initial
+          # Install gdal using pip3 into the Poetry virtualenv
+          poetry run pip3 install gdal==3.0.4
+          # Install rest of project          
+          poetry install --without test,black
```
[the remark this answers is not in the document]
Comment (rod-glover) [reply .github/workflows/pypi-publish.yml:48]: Whoa, I didn't know you could build without installing. That's great.

File: pdp_util/pcds_index.py [hunk 6]
```diff
@@ -176,41 +177,24 @@ def get_elements(self, sesh):
...
+            .join(VarsPerHistory, VarsPerHistory.history_id == History.id)
+            .join(Variable, Variable.id == VarsPerHistory.vars_id)
+            # .join(Variable, Variable.network_id == Network.id)
+            .filter(Network.name == network_name)
+            .distinct()
+            .order_by(Station.native_id)
+        )
+        climo_filter = or_(
+            Variable.cell_method.contains("within"),
+            Variable.cell_method.contains("over"), )
```
[the remark this answers is not in the document]
Comment (rod-glover) [reply pdp_util/pcds_index.py:194]: Yes. The transition to Poetry is the setup for that PR. But it requires fairly far-ranging work, definitely out of scope for this PR.

File: .github/workflows/black.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (rod-glover) [reply .github/workflows/black.yml:11]: Better yet, I have temporarily disabled the workflow through GitHub. :man_facepalming:

File: .github/workflows/pypi-publish.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (rod-glover) [reply .github/workflows/pypi-publish.yml:48]: My experimentation indicates that you don't even need to create a virtual env:

```
(pdp-util-py3.8) rglover@pcic-3002:~/code$ git clone git@github.com:pacificclimate/pdp_util.git pdp_util2
Cloning into 'pdp_util2'...
remote: Enumerating objects: 1517, done.
remote: Counting objects: 100% (419/419), done.
remote: Compressing objects: 100% (150/150), done.
remote: Total 1517 (delta 307), reused 331 (delta 258), pack-reused 1098
Receiving objects: 100% (1517/1517), 463.68 KiB | 770.00 KiB/s, done.
Resolving deltas: 100% (977/977), done.
(pdp-util-py3.8) rglover@pcic-3002:~/code$ cd pdp_util2
(pdp-util-py3.8) rglover@pcic-3002:~/code/pdp_util2$ git checkout i-51-Poetry
Branch 'i-51-Poetry' set up to track remote branch 'i-51-Poetry' from 'origin'.
Switched to a new branch 'i-51-Poetry'
(pdp-util-py3.8) rglover@pcic-3002:~/code/pdp_util2$ poetry build
Building pdp-util (2.1.0)
  - Building sdist
  - Built pdp_util-2.1.0.tar.gz
  - Building wheel
  - Built pdp_util-2.1.0-py3-none-any.whl
```

[the remark this answers is not in the document]
Comment (rod-glover) [reply .github/workflows/pypi-publish.yml:48]: Seems you may not need to explicitly create the venv, even:

```
rglover@pcic-3002:~/code$ git clone git@github.com:pacificclimate/pdp_util.git pdp_util2
Cloning into 'pdp_util2'...

rglover@pcic-3002:~/code$ cd pdp_util2

rglover@pcic-3002:~/code/pdp_util2$ git checkout i-51-Poetry
Branch 'i-51-Poetry' set up to track remote branch 'i-51-Poetry' from 'origin'.
Switched to a new branch 'i-51-Poetry'

rglover@pcic-3002:~/code/pdp_util2$ poetry build 
Creating virtualenv pdp-util-cs7JC6yA-py3.8 in /home/rglover/.cache/pypoetry/virtualenvs
Building pdp-util (2.1.0)
```

File: .github/workflows/python-ci.yml [hunk 7]
```diff
@@ -8,44 +8,55 @@ jobs:
     runs-on: ubuntu-20.04
     strategy:
       matrix:
-        python-version: [3.7, 3.8]
+        python-version:
+          - "3.8"
+          # TODO: Test with other Pythons
```
[the remark this answers is not in the document]
Comment (rod-glover) [reply .github/workflows/python-ci.yml]: Yes, I did a few tries with 3.9 - 3.11 and realized there was a lot of work to do there.
